[FLINK-6830] Add StatefulJobSavepointFrom13MigrationITCase

tzulitai opened the pull request:
With [FLINK-6763](https://issues.apache.org/jira/browse/FLINK-6763) and [FLINK-6764](https://issues.apache.org/jira/browse/FLINK-6764) we'll need to change the serialization formats between 1.3.0 and 1.3.x.

This PR adds the ITCase for the savepoint migration so that those format changes will be properly guarded for backwards compatibility.

The savepoint binary files of this PR were taken using the current `release-1.3` branch.

zentol commented:
Can we also include the tests for the restore with topology changes in flink-tests?

org.apache.flink.test.start.operator.restore *

aljoscha commented:
The changes look good! @zentol's suggestion also seems wise. 👌

tzulitai commented:
Thanks for the reviews @zentol @aljoscha! I'll add the topology change ITTests also and then merge this.

aljoscha commented:
@tzulitai While you're on it, what would you think about also porting the `*From12MigrationTest` tests?

tzulitai commented:
@aljoscha +1 to that, I'm actually working on those also already :-D

tzulitai commented:
f228081 is for porting the topology change on restore migration tests.

The rest 233c617 to e619b98 is for ports of each individual migration test with the pattern `*From12MigrationTest`. They're also refactored to deduplicate code across 1.1, 1.2, 1.3 migration tests.

Note that for the CEP migration tests in 233c617, some tests for migration from 1.3 is currently ignored because they do not pass. May have bumped into a bug for the CEP library that does not allow it to restore from 1.3.x to 1.3.x 😅

zentol commented on flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/unkeyed/ChainBreakTest.java:
```diff
@@ -31,8 +36,22 @@
+	private final String savepointPath;
+
+	@Parameterized.Parameters(name = "Migrate Savepoint: {0}")
+	public static Collection<String> parameters () {
```
is it not possible to move this into the abstract super class?

tzulitai replied:
Will change!

zentol commented:
topology tests look good.

aljoscha commented:
Why are the CEP "from12" snapshots also changed?

aljoscha commented:
@kl0u Could you please look into the CEP tests that failing to restore from 1.3 to 1.3?

kl0u commented:
@aljoscha yes we have already discussed offline with @tzulitai .

aljoscha commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/WindowOperatorMigrationTest.java:
```diff
@@ -149,9 +169,16 @@ public void testRestoreSessionWindowsWithCountTrigger() throws Exception {
-						OperatorSnapshotUtil.getResourceFilename("win-op-migration-test-session-with-stateful-trigger-flink1.2-snapshot")));
+
+		String savepointFile = "win-op-migration-test-session-with-stateful-trigger-flink" + testMigrateVersion + "-snapshot";
+		if (testMigrateVersion.equals("1.1")) {
```
I think you could refactor this piece of code into a method, something like:
```
restoreFromSnapshot(TestHarness, SnapshotPath, MigrationVersion)
```

What do you think?

tzulitai replied:
That makes sense! I'll add the method to some `MigrationTestUtil` perhaps.

aljoscha commented:
I like these changes a lot! Especially the refactoring of using the same test for all versions. (I'm sure @zentol will also like that 😉 )

Had some inline comments.

kl0u commented:
Hi @tzulitai ! Thanks for testing also this. 

The problem with the `CEP` failing test for `1.3` while passing for `1.2` is that the semantics of the `followedBy()` changed between `1.2` and `1.3`, so the expected results cannot be the same. 

In `1.2`, `followedBy()` implied non-deterministic relaxed contiguity while in `1.3` it is just relaxed contiguity. So for `1.3`, either change the expected result to contain only one match with the `foo1` as middle element, or change the pattern to have `followedByAny(middle)` (and in this case you have to take the savepoint again).

tzulitai commented:
Thanks @kl0u, your suggestions fixes the problem. Would it also be possible to include document the contiguity semantic changes in #4041?

kl0u commented:
@tzulitai yes I will ;)

tzulitai commented:
Addressed @aljoscha @kl0u @zentol's comments.

Also rebased on `master` for another Travis run. If green, will merge by the end of today to `master` and `release-1.3` if there's no other objections!

tzulitai commented:
Merging ...
